myia/anf_ir.py

```diff
@@ -21,6 +21,31 @@
 APPLY = Named('APPLY')
 
 
+class Debug(types.SimpleNamespace):
+    """Debug information for an object.
+
+    Attributes:
+        name: The name of the object.
+
+    """
+
+    _curr_id = 0
+
+    def __init__(self, **kwargs):
+        """Construct a Debug object."""
+        self.name: str = None
+        super().__init__(**kwargs)
+
+    @property
+    def debug_name(self):
+        """Return the name, create a fresh name if needed."""
+        if self.name:
+            return self.name
+        Debug._curr_id += 1
+        self.name = f'#{Debug._curr_id}'
```

Comment:
Too late, but `#1` in error messages would interfere with GitHub issues if they ever get reported by users.

Reply:
I actually changed it in my local branch because #1 is also not a valid variable name, which gets a bit awkward in some object representations. We can change it.

Reply:
It shouldn't be a valid variable name: this lets us have terse names that nonetheless can never clash with the user's choices. Could be `$1`, though.

Reply:
Don't we only generate names if they don't already exist? So how could it clash with a user's choice?

Why should it not be a valid variable name? For example, the `__repr__` of a node should probably use the `__str__` representation of its inputs (otherwise you end up printing the entire graph recursively). You then end up with things like `Apply(inputs=[#1, #2])`, which I think looks weird.

Reply:
I mean it could clash with the names of other nodes. If e.g. we generate the name `arg0`, the user could have used that name elsewhere.

> You then end up with things like `Apply(inputs=[#1, #2])`, which I think looks weird.

Personally I think it looks weird in a good way, it tells me that these are automatically named temporaries better than valid variable names would. The "best" alternative would be `_1`, `_2`, etc. but I think that's just ugly.

Reply:
Let's continue discussing in #17.